doc/man/gnunet-tracekit.1
.TH GNUNET-TRACEKIT "1" "20 Feb 2003" "GNUnet"
.SH NAME
gnunet-tracekit \- overlay network traceing
.SH SYNOPSIS
.B gnunet\-tracekit
[\fIOPTIONS\fR] 
.SH DESCRIPTION
gnunet-tracekit can be used to visualize the GNUnet overlay
network.  This tool broadcasts a message into the network
querying every peer about a list of other connected peers.
The result is then printed to the screen.
.PP
gnunet-tracekit is primarily useful for developers to determine
the structure of the network.  End-users will find little use
for the tool itself (except to satisfy their curiosity) but
should at the moment still enable this protocol to help the
developers that trace the network.
.PP
.TP
\fB\-h\fR, \fB\-\-help\fR
print help page
.TP
\fB\-c \fIFILENAME\fR, \fB\-\-config=\fIFILENAME\fR
load config file (defaults: ~/.gnunet/gnunet.conf)
\fB\-W \fIWAIT\fR, \fB\-\-wait=\fIWAIT\fR
set how many seconds to wait for replies
\fB\-D \fIDEPTH\fR, \fB\-\-depth=\fIDEPTH\fR
until which depth should the network be traced (how many hops)
\fB\-F \fIFORMAT\fR, \fB\-\-format=\fIFORMAT\fR
specifies the output format, 0 is human readable, 1 is dot
\fB\-P \fIPRIORITY\fR, \fB\-\-priority=\fIPRIORITY\fR
what should the priority of the probe be
.TP
\fB\-v\fR, \fB\-\-version\fR
print the version number
.TP
\fB\-L \fILOGLEVEL\fR, \fB\-\-loglevel=\fILOGLEVEL\fR
change the loglevel. Possible values for \fILOGLEVEL\fR are NOTHING, FATAL, ERROR, FAILURE, WARNING, MESSAGE, INFO, DEBUG, CRON and EVERYTHING.
.SH NOTES
gnunet-tracekit may not be supported by all GNUnet peers.  Peers that
do not support tracekit will still be discovered and will be marked specially
in the output.
.SH FILES
.TP
~/.gnunet/gnunet.conf
Users GNUnet configuration file
.SH "REPORTING BUGS"
Report bugs by using mantis <http://www.ovmj.org/~mantis/> or by sending electronic mail to <gnunet-developers@gnu.org>
.SH "SEE ALSO"
\fBgnunet.conf\fP(5), \fBgnunetd\fP(1)


ChangeLog
Mon Feb 28 00:33:03 EST 2005
	Releasing GNUnet 0.6.6b.

Sun Feb 27 19:27:44 EST 2005
	Fixing mantis 839.

Sun Feb 27 18:44:11 EST 2005
	Fixing mantis 833/834.

Sun Feb 27 18:15:02 EST 2005
	Fixing serious problem with LE in gnunet-insert.
	Fixing mantis 828.

Thu Feb 17 01:14:30 EST 2005
	Various memory leaks for Win32 fixed.
	Added support for GNUNET(D)_CONFIG environment variable.

Wed Feb  9 13:33:38 EST 2005
        Fixed problems with the logrotate code.  Releasing GNUnet 0.6.6a.

Sun Jan 11 19:58:13 CET 2005
	Releasing GNUnet 0.6.6

Sun Jan 11 19:24:27 CET 2005
        gnunet-gtk now resumes aborted downloads automatically.

Sun Nov 14 16:49:02 EST 2004
	Debugged collections, seems to work (not that it looks pretty,
	but it basically works, the rest is UI work).  Fixed deadlock
	in gnunet-gtk logging.  Releasing GNUnet 0.6.5.

Thu Nov  4 09:04:48 EST 2004
	Added (highly experimental) support for collections.

Sat Sep 25 22:29:48 EST 2004
	Releasing GNUnet 0.6.4a.

Sun Sep 19 23:39:04 EST 2004
	Various code cleanups and minor bugfixes.

Sun Aug 29 15:11:28 IST 2004
	Releasing GNUnet 0.6.4.

Tue Aug 24 20:59:14 IST 2004
	Significant enhancements to namespaces.  A namespace can
	now be annotated with meta-data at the time that it is
	created.
	
Sat Aug 21 01:00:06 IST 2004
	Adding GNU gettext support to configure.
	
Thu Aug 19 01:26:55 IST 2004
	Migrating gnunet-gtk to GTK 2.0 (tested with 2.4).

Mon Aug 16 00:30:59 IST 2004
	Removed gcry.  Added dependency on GNU MP library (libgmp)
	for kblocks.

Tue Aug  3 17:57:43 IST 2004
	Fixed bugs in gnunet-update.  Fixed segfault in AFS startup.
	Fixed warnings in gnunet-peer-info/gnunetd.  Fixed packaging
	error (not all of util/gcry included in distribution).
	Releasing GNUnet 0.6.3a.

Mon Aug  2 20:06:07 IST 2004
	Releasing GNUnet 0.6.3.

Sun Aug  1 01:31:19 IST 2004
	Changed most places from HexName to the new, shorter EncName.
	Pushed URIs for namespaces and keyword searches into more
	widespread use.  Replaced all occurences of sprintf with
	SNPRINTF (to obtain extra bounds check).
	
Sun Jul 11 13:54:01 IST 2004
	Moved testcases for libgnunet_util to util/ directory.
	Made DB testcases in afs/module/ run for each available
	database and not just the one configured in /etc/gnunet.conf.

Fri Jun 25 21:05:59 IST 2004
	Releasing GNUnet 0.6.2b (resolves various minor bugs).

Wed May  5 17:34:39 EST 2004
	Releasing GNUnet 0.6.2a (resolves libgcrypt and packageing
	issues, improved gtk-statistics, other minor bugfixes).
	
Fri Apr 30 21:36:23 EST 2004
	Fixed synchronization problem in cron (only relevant for
	parallel downloading).  Releasing GNUnet 0.6.2.

Mon Apr 26 21:13:45 EST 2004
	Added support to gnunet-check to convert file index database
	to new design.  Removed support for pre 0.6.1b bloomfilter
	conversion (would have been too messy to support both).
	Added old-version check to gnunetd startup to make updates
	more systematic in the future (not perfect, though).

Mon Apr 26 19:27:29 EST 2004
	Recursive insertion of directories with gnunet-gtk 
	now works including the progress bar.

Sun Apr 25 18:24:55 EST 2004
	Global changes to fix bug #698.  The fix is still 
	incomplete with respect to pthread.h specific structs.
	
Sun Apr 25 15:20:29 EST 2004
	Fixed compile error in IPv6 transports.

Fri Apr 23 23:38:01 EST 2004
	Added recursive insertion of directories to gnunet-gtk.
	Progress-bar does not work yet, also libextractor is
	always used (no option to turn it off).
	
Fri Apr 23 21:49:18 EST 2004
	Removed need to specify _protocol in configuration file
	(shorter is better).

Wed Apr 21 23:32:36 EST 2004
	Added code to bound priority used by clients by twice the
	average priority of requests from other peers that are 
	routed at the moment.  That prevents clients from exposing
	themselves with excessively high priorities (and also 
	avoids giving too much credit/trust needlessly).

Tue Apr 20 22:27:19 EST 2004
	Added possibility of using a link instead of making a copy
	if gnunetd and the insert-client both run on the same machine.

Tue Apr 20 00:00:58 EST 2004
	Added network transparency for indexing files (by copying
	them to a GNUnet(d) specific directory).  This breaks
	unindexing (for files indexed with previous versions) and
	forces the user to make a copy of the file.  In the near
	future we should resolve this problem by allowing the use
	of a link.

Sun Apr 18 17:24:34 EST 2004
	Adding recursive, parallel download of directories to
	gnunet-download

Sun Apr 18 01:57:32 EST 2004
	gnunet-transport-check can now be used to check
	external connectivity (with other peers).
	
Sat Apr 17 17:46:19 EST 2004
	Reduced memory consuption for AFS insertion and
	download.
	
Sat Apr 10 21:17:46 EST 2004
	Implemented statistics-plot in gnunet-gtk based on
	gnome-system-monitor code.

Fri Apr  9 22:01:51 EST 2004
	Added support for libextractor > 0.2.7.
	
Fri Apr  9  0:29:20 EST 2004
	Completed --help conversion.

Wed Mar 31 15:40:51 EST 2004
	Releasing 0.6.1d.

Tue Mar 30 22:42:13 EST 2004
	Integrated gnunet-pseudonym-create, -delete and -list into
	gnunet-pseudonym as well as gnunet-directory-list, -emptydb
	and -print into gnunet-directory.

Tue Mar  2 10:46:01 EST 2004
	Fixing bug preventing download for indexed files <= 1k.

Sat Feb 28 03:43:34 EST 2004
	Releasing 0.6.1c.

Sat Feb 21 06:16:47 EST 2004
	Modified requestmanager to improve performance at the end of the 
	download.

Thu Feb 19 00:48:33 EST 2004
	Fixed integer underflow possibly causing slow download performance.
	
Sun Feb 15 15:17:05 EST 2004
	Fixed weak key failures with libgcrypt.

Sun Feb 15 13:14:44 EST 2004
	New routing code (untested) commited.

Sun Feb 15 11:58:10 EST 2004
	Fixing problem with 64-bit size_t on alpha (#655).

Sat Jan 31 04:32:33 EST 2004
	Releasing 0.6.1b.

Fri Jan 23 00:00:09 EST 2004
	Fixed indexed content being overwritten by migrated content
	
Thu Jan 22 19:27:33 EST 2004
	Fixed integer-overflow problems in statuscalls.

src/applications/tracekit/gnunet-tracekit.c
/*
     This file is part of GNUnet.
     (C) 2001, 2002, 2004 Christian Grothoff (and other contributing authors)

     GNUnet is free software; you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published
     by the Free Software Foundation; either version 2, or (at your
     option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     General Public License for more details.

     You should have received a copy of the GNU General Public License
     along with GNUnet; see the file COPYING.  If not, write to the
     Free Software Foundation, Inc., 59 Temple Place - Suite 330,
     Boston, MA 02111-1307, USA.
*/

/**
 * @file applications/tracekit/gnunet-tracekit.c 
 * @brief tool that sends a trace request and prints the received network topology
 * @author Christian Grothoff
 */

#include "tracekit.h"
#include "platform.h"

#define TRACEKIT_VERSION "0.0.3"

static Semaphore * doneSem;

/**
 * Parse the options, set the timeout.
 * @param argc the number of options
 * @param argv the option list (including keywords)
 * @return OK on error, SYSERR if we should exit 
 */
static int parseOptions(int argc,
			char ** argv) {
  int option_index;
  int c;  

  FREENONNULL(setConfigurationString("GNUNETD",
				     "LOGFILE",
				     NULL));
  while (1) {
    static struct GNoption long_options[] = {
      LONG_DEFAULT_OPTIONS,
      { "depth", 1, 0, 'D' },
      { "format", 1, 0, 'F' },
      { "priority", 1, 0, 'P' },
      { "wait", 1, 0, 'W' },
      { 0,0,0,0 }
    };    
    option_index = 0;
    c = GNgetopt_long(argc,
		      argv, 
		      "vhdc:L:H:W:D:F:P:", 
		      long_options, 
		      &option_index);    
    if (c == -1) 
      break;  /* No more flags to process */
    if (YES == parseDefaultOptions(c, GNoptarg))
      continue;
    switch(c) {
    case 'D': {
      unsigned int depth;
      if (1 != sscanf(GNoptarg, "%ud", &depth)) {
	LOG(LOG_FAILURE, 
	    _("You must pass a number to the '%s' option.\n"),
	    "-D");
	return SYSERR;
      } else {
	setConfigurationInt("GNUNET-TRACEKIT",
			    "HOPS",
			    depth);
      }
      break;
    }
    case 'F': {
      unsigned int format;
      if (1 != sscanf(GNoptarg, "%ud", &format)) {
	LOG(LOG_FAILURE, 
	    _("You must pass a number to the '%s' option.\n"),
	    "-F");
	return SYSERR;
      } else {
	setConfigurationInt("GNUNET-TRACEKIT",
			    "FORMAT",
			    format);
      }
      break;
    }
    case 'h': {
      static Help help[] = {
	HELP_CONFIG,
	{ 'D', "depth", "DEPTH",
	  gettext_noop("probe network to the given DEPTH") },
	{ 'F', "format", "FORMAT",
	  gettext_noop("specify output format; 0 for human readable output, 1 for dot, 2 for vcg") },
	HELP_HELP,
	HELP_LOGLEVEL,
	{ 'P', "priority", "PRIO",
	  gettext_noop("use PRIO for the priority of the trace request") },
	HELP_VERSION,
	{ 'W', "wait", "DELAY",
	  gettext_noop("wait DELAY seconds for replies") },
	HELP_END,
      };
      formatHelp("gnunet-tracekit [OPTIONS]",
		 _("Trace GNUnet network topology."),
		 help);
      return SYSERR;
    }
    case 'P': {
      unsigned int prio;
      if (1 != sscanf(GNoptarg, "%ud", &prio)) {
	LOG(LOG_FAILURE, 
	    _("You must pass a number to the '%s' option.\n"),
	    "-P");
	return SYSERR;
      } else {
	setConfigurationInt("GNUNET-TRACEKIT",
			    "PRIORITY",
			    prio);
      }
      break;
    }
    case 'W': {
      unsigned int wait;
      if (1 != sscanf(GNoptarg, "%ud", &wait)) {
	LOG(LOG_FAILURE, 
	    _("You must pass a number to the '%s' option.\n"),
	    "-W");
	return SYSERR;
      } else {
	setConfigurationInt("GNUNET-TRACEKIT",
			    "WAIT",
			    wait);
      }
      break;
    }
    case 'v': 
      printf("GNUnet v%s, gnunet-tracekit v%s\n",
	     VERSION,
	     TRACEKIT_VERSION);
      return SYSERR;
    default: 
      LOG(LOG_FAILURE,
	  _("Use --help to get a list of options.\n"));
      return SYSERR;
    } /* end of parsing commandline */
  } /* while (1) */
  return OK;
}

static void * receiveThread(GNUNET_TCP_SOCKET * sock) {
  TRACEKIT_CS_REPLY * buffer;
  int format;
  HostIdentity * peersSeen;
  int psCount;
  int psSize;
  HostIdentity * peersResponding;
  int prCount;
  int prSize;
  int i;
  int j;
  int match;

  psCount = 0;
  psSize = 1;
  peersSeen = MALLOC(psSize * sizeof(HostIdentity));
  prCount = 0;
  prSize = 1;
  peersResponding = MALLOC(prSize * sizeof(HostIdentity));
  buffer = MALLOC(MAX_BUFFER_SIZE);
  format = getConfigurationInt("GNUNET-TRACEKIT", 
			       "FORMAT");
  if (format == 1)
    printf("digraph G {\n");
  if (format == 2)
    printf("graph: {\n");
  while (OK == readFromSocket(sock, 
			      (CS_HEADER**)&buffer)) {
    int count;
    EncName enc;

    count = ntohs(buffer->header.size) - sizeof(TRACEKIT_CS_REPLY);
    if (count < 0) {
      BREAK();
      break; /* faulty reply */
    }
    hash2enc(&buffer->responderId.hashPubKey,
	     &enc);
    match = NO;
    for (j=0;j<prCount;j++)
      if (equalsHashCode160(&buffer->responderId.hashPubKey,
			    &peersResponding[j].hashPubKey))
	match = YES;
    if (match == NO) {
      if (prCount == prSize) 
	GROW(peersResponding,
	     prSize,
	     prSize*2);
      memcpy(&peersResponding[prCount++],
	     &buffer->responderId.hashPubKey,
	     sizeof(HostIdentity));
    }
    count = count / sizeof(HostIdentity);
    if (ntohs(buffer->header.size) !=
	sizeof(TRACEKIT_CS_REPLY) +
	count * sizeof(HostIdentity)) {
      BREAK();
      break;
    }
    if (count == 0) {
      switch (format) {
      case 0:
	printf(_("'%s' is not connected to any peer.\n"),
	       (char*)&enc);
	break;
      case 1:
	printf("  %.*s;\n",
	       4, (char*)&enc);
	break;
      case 2:
	/* deferred -- vcg needs all node data in one line */
	break;
      default:
	printf(_("Format specification invalid. "
		 "Use 0 for user-readable, 1 for dot, 2 for vcg.\n"));
	break;
      }
    } else {
      EncName other;

      for (i=0;i<count;i++) {
	match = NO;
	for (j=0;j<psCount;j++)
	  if (equalsHashCode160(&((TRACEKIT_CS_REPLY_GENERIC*)buffer)->peerList[i].hashPubKey,
				&peersSeen[j].hashPubKey))
	    match = YES;
	if (match == NO) {
	  if (psCount == psSize)
	    GROW(peersSeen,
		 psSize,
		 psSize * 2);
	  memcpy(&peersSeen[psCount++],
		 &((TRACEKIT_CS_REPLY_GENERIC*)buffer)->peerList[i].hashPubKey,
		 sizeof(HostIdentity));
	}

	hash2enc(&((TRACEKIT_CS_REPLY_GENERIC*)buffer)->peerList[i].hashPubKey,
		 &other);
	switch (format) {
	case 0:
	  printf(_("'%s' connected to '%s'.\n"),
		 (char*)&enc,
		 (char*)&other);
	  break;
	case 1: /* dot */
	  printf("  \"%.*s\" -> \"%.*s\";\n",
		 4, (char*)&enc,
		 4, (char*)&other);
	  break;
	case 2: /* vcg */
	  printf("\tedge: { sourcename: \"%s\" targetname: \"%s\" }\n",
		 (char*)&enc,
		 (char*)&other);
	  break;
	default: /* undef */
	  printf(_("Format specification invalid. "
		   "Use 0 for user-readable, 1 for dot\n"));
	  break;
	}
      }
    }
  }
  FREE(buffer);
  for (i=0;i<psCount;i++) {
    EncName enc;

    match = NO;
    for (j=0;j<prCount;j++)
      if (equalsHashCode160(&peersResponding[j].hashPubKey,
			    &peersSeen[i].hashPubKey)) {
	match = YES;
	break;
      }
    if (match == NO) {
      hash2enc(&peersSeen[i].hashPubKey,
	       &enc);
      switch (format) {
      case 0:
	printf(_("Peer '%s' did not report back.\n"),
	       (char*)&enc);
	break;
      case 1:
	printf("  \"%.*s\" [style=filled,color=\".7 .3 1.0\"];\n",
	       4,
	       (char*)&enc);
	break;
      case 2:
	printf("\tnode: { title: \"%s\" label: \"%.*s\" shape: \"ellipse\" }\n",
	       (char*)&enc,
	       4, (char*) &enc);
	break;
      default:
	break;
      }
    } else {
      switch (format) {
      case 2:
	hash2enc(&peersSeen[i].hashPubKey,
		 &enc);
	printf("\tnode: { title: \"%s\" label: \"%.*s\" }\n",
	       (char*)&enc, 4, (char*)&enc);
	break;
      default:
	break;
      }
    }
  }
  if (psCount == 0) {
    switch (format) {
    case 2:
      printf("\tnode: { title: \"NO CONNECTIONS\" }\n");
      break;
    default:
      break;
    }
  }
  if (format == 1)
    printf("}\n");
  if (format == 2)
    printf("}\n");
  SEMAPHORE_UP(doneSem);
  FREE(peersResponding);
  FREE(peersSeen);  
  return NULL;
}

/**
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return return value from gnunet-tracekit: 0: ok, -1: error
 */   
int main(int argc, char ** argv) {
  GNUNET_TCP_SOCKET * sock;
  PTHREAD_T messageReceiveThread;
  void * unused;
  TRACEKIT_CS_PROBE probe;
  int sleepTime;

  if (SYSERR == initUtil(argc, argv, &parseOptions))
    return 0; /* parse error, --help, etc. */ 
  sock = getClientSocket();
  if (sock == NULL) {
    LOG(LOG_ERROR,
	_("Could not connect to gnunetd.\n"));
    return -1;
  }

  doneSem = SEMAPHORE_NEW(0);
  if (0 != PTHREAD_CREATE(&messageReceiveThread, 
			  (PThreadMain) &receiveThread, 
			  sock,
			  128 * 1024))
    DIE_STRERROR("pthread_create");
  
  probe.header.size 
    = htons(sizeof(TRACEKIT_CS_PROBE));
  probe.header.tcpType
    = htons(TRACEKIT_CS_PROTO_PROBE);
  probe.hops 
    = htonl(getConfigurationInt("GNUNET-TRACEKIT",
				"HOPS"));
  probe.priority
    = htonl(getConfigurationInt("GNUNET-TRACEKIT",
				"PRIORITY"));  
  if (SYSERR == writeToSocket(sock,
                              &probe.header)) {
    LOG(LOG_ERROR,
	_("Could not send request to gnunetd.\n"));
    return -1;
  }
  startCron();
  initializeShutdownHandlers();
  sleepTime = getConfigurationInt("GNUNET-TRACEKIT",
                                  "WAIT");
  if (sleepTime == 0)
    sleepTime = 5;
  addCronJob((CronJob)&run_shutdown,
	     cronSECONDS * sleepTime,
	     0,
	     NULL);
  wait_for_shutdown();
  closeSocketTemporarily(sock);
  SEMAPHORE_DOWN(doneSem);
  SEMAPHORE_FREE(doneSem);
  PTHREAD_JOIN(&messageReceiveThread, &unused);
  doneShutdownHandlers();
  releaseClientSocket(sock);
  stopCron();
  doneUtil();
  return 0;
}

/* end of gnunet-tracekit.c */
